docs/src/developer/developer/cassandra-interaction.md

```diff
@@ -1,46 +1,301 @@
-    * [Backwards incompatible schema changes](#backwards-incompatible-schema-changes)
-        * [What to do about backwards incompatible schema changes](#what-to-do-about-backwards-incompatible-schema-changes)
+Cassandra was chosen back in ~2014 in the context of massive scalability (expectations of billions
+of users), as well as the wish for redundancy and cloud-always-on high-availability.
 
-<!-- vim-markdown-toc -->
+From an operational standpoint, high-availability and redundancy are well served with this: any
+single node/VM can crash at any moment without data loss, and version upgrades, server maintenance,
+and server migrations can be done without stopping the system (and causing user-observable
+downtime).
 
-## Anti-patterns
+From a developer perspective, wishing for a mental model of strong consistency across multiple
+tables in multiple services to make easier assumptions when writing code makes working with
+cassandra sometimes a bit harder. (That said, this document has some sections below to make working
+with strong consistency within the same service easier).
 
-### Anti-pattern: Using full table scans in production code
+Questions have come up whether another technology may serve our needs better in the evolved
+technical and product context of 2023 and beyond. These questions are valid, and a change can always
+be discussed, if the time is at hand to put this effort in.
+
+## Guidelines for designing database schemas
+
+In a lot of relational databases, you design tables with their relations, and think about the exact
+access patterns later, as you can always add an index on a certain column. In the cassandra world,
+you need to make some mental gymnastics and get used to design your tables **by the access patterns
+to this data**. You can only look up data by what's in the primary key, and you always need the
+single partition key at hand to do this lookup.
+
+Therefore, you may need multiple tables that relate, such as a `conversation_by_user_id` and a
+`user_by_conversation_id` table (or even more of them). That's perfectly fine and the correct
+pattern to use. Please read some online resources about data modeling with cassandra, such as the
+[datastax official data modelling documentation](https://cassandra.apache.org/doc/latest/cassandra/data_modeling/index.html).
+
+## Guidelines on per-table query consistency level and replication level
+
+### Background
+
+We currently operate all our production and production-like environments with
+
+* a **replication factor of 3** (per datacentre, but except for a brief phase of controlled
+  datacentre migration there is only one datacentre in use)
+* a number of cassandra nodes >= 3 (depending on load and disk/machine size)
+* Everywhere where some consistency matters, **Quorum writes and Quorum reads**
+
+These three things together give us **strong consistency on a per-table** basis. A quorum write
```

Comment:
> > Then Y reads from A, B and sees the new value
> 
> Ahm, I understand this differently:

Ah, yes, my example is wrong. There are a few problems with your statements below, though:

> When Y reads from A and B - and the new value has only been written to A - then the quorum read will not agree (value on A and on B are different) so then Y will wait for an answer from C as well, to have a quroum - B and C agreeing on the value.

This is not correct. What happens is that after receiving an answer from A and B, Cassandra will compare timestamps and return the new value to the client (so the one coming from A), without waiting for C.
Meanwhile, Z might be getting values from B and C instead, and return the old value. So this part of the example is correct.

My mistake was thinking that this violates strong consistency. It actually doesn't. There is still an order of operations that makes sense (X, then Z, then Y).

There is a subtlety here I was confused about. Suppose it's Y itself that performs the second read. Then you might think you could witness a violation of monotonic read consistency, because Y would receive the old value the second time. But in fact, that's not the case, because Cassandra employs a "read repair" mechanism, that reconciles B with the new value when it's part of the quorum. This would happen on the first read, in the example.

> But perhaps the example is slightly weirdly chosen, the point you make is correct. At a given time t, one process can read the old value and another process can read the new value. Until the quorum write completes - at that point the old value cannot be read anymore by any process. But! even with a full ACID database, if you have a bunch of parallel requests reading and writing at roughly the same time, then all those processes reading before the write happens see an old value, and all those reading after the write happens see the new value.

This would be strongly consistent, I think you're confusing strong consistency with strict consistency. For example strong consistency doesn't imply that executions are deterministic.